unlock_indicator.c

```diff
@@ -85,6 +87,48 @@ static xcb_visualtype_t *vistype;
 unlock_state_t unlock_state;
 auth_state_t auth_state;
 
+static void string_append(char** string_ptr, const char* appended) {
+    char *tmp = NULL;
+    if (*string_ptr == NULL) {
+        if (asprintf(&tmp, "%s", appended) != -1) {
+            *string_ptr = tmp;
+        }
+    } else if (asprintf(&tmp, "%s, %s", *string_ptr, appended) != -1) {
+        free(*string_ptr);
+        *string_ptr = tmp;
+    }
+}
+
+static void display_button_text(cairo_t *ctx, const char* text, double y_offset) {
+    cairo_text_extents_t extents;
+    double x, y;
+
+    cairo_text_extents(ctx, text, &extents);
+    x = BUTTON_CENTER - ((extents.width / 2) + extents.x_bearing);
+    y = BUTTON_CENTER - ((extents.height / 2) + extents.y_bearing) + y_offset;
+
+    cairo_move_to(ctx, x, y);
+    cairo_show_text(ctx, text);
+    cairo_close_path(ctx);
+}
+
+static void update_layout_string() {
+    if (layout_string) {
+        free(layout_string);
+        layout_string = NULL;
+    }
+    xkb_layout_index_t num_layouts = xkb_keymap_num_layouts(xkb_keymap);
+    for (xkb_layout_index_t i = 0; i < num_layouts; ++i) {
+        if (xkb_state_layout_index_is_active(xkb_state, i, XKB_STATE_LAYOUT_EFFECTIVE)) {
+            const char* name = xkb_keymap_layout_get_name(xkb_keymap, i);
+            if (!name) {
+                name = "Unknown";
```

Comment:
It would probably be better to display nothing rather than “Unknown”, as that doesn’t convey to the user what’s missing (but “keyboard layout: unknown” is too long).

Reply:
Done